blocks/rich-text/format.js

```diff
@@ -0,0 +1,102 @@
+/**
+ * External dependencies
+ */
+import { omitBy, map } from 'lodash';
+import { nodeListToReact } from 'dom-react';
+
+/**
+ * WordPress dependencies
+ */
+import { createElement, renderToString } from '@wordpress/element';
+
+/**
+ * Transforms a WP Element to its corresponding HTML string.
+ *
+ * @param {WPElement} value Element.
+ *
+ * @return {string} HTML.
+ */
+export function elementToString( value ) {
+	return renderToString( value );
+}
+
+/**
+ * Transforms a value in a given format into string.
+ *
+ * @param {Array|string?}  value  DOM Elements.
+ * @param {string} format Output format (string or element)
+ *
+ * @return {string} HTML output as string.
+ */
+export function valueToString( value, format ) {
+	switch ( format ) {
+		case 'string':
+			return value || '';
+		default:
+			return elementToString( value );
+	}
+}
+
+/**
+ * Strips out TinyMCE specific attributes and nodes from a WPElement
+ *
+ * @param {string} type    Element type
+ * @param {Object} props   Element Props
+ * @param {Array} children Element Children
+ *
+ * @return {Element} WPElement.
+ */
+export function createTinyMCEElement( type, props, ...children ) {
+	if ( props[ 'data-mce-bogus' ] === 'all' ) {
+		return null;
+	}
+
+	if ( props.hasOwnProperty( 'data-mce-bogus' ) ) {
+		return children;
+	}
+
+	return createElement(
+		type,
+		omitBy( props, ( _, key ) => key.indexOf( 'data-mce-' ) === 0 ),
+		...children
+	);
+}
+
+/**
+ * Transforms an array of DOM Elements to their corresponding WP element.
+ *
+ * @param {Array} value DOM Elements.
+ *
+ * @return {WPElement} WP Element.
+ */
+export function domToElement( value ) {
+	return nodeListToReact( value || [], createTinyMCEElement );
+}
+
+/**
+ * Transforms an array of DOM Elements to their corresponding HTML string output.
+ *
+ * @param {Array} value DOM Elements.
+ *
+ * @return {string} HTML.
+ */
+export function domToString( value ) {
+	return map( value, element => element.outerHTML ).join( '' );
```

Comment:
In `domToElement`, we're always stripping any TinyMCE internal element and attributes, but here we're not, even though there might be cases where we receive some in the case of splitting.

Reply:
Not sure how to address this in an optimal way. Don't we have tools in TinyMCE to do this for us instead of adding custom logic? Is it a requirement since we're not using this when getting the whole content of the editor but only when splitting?

Reply:
`( new tinymce.dom.Serializer( {} ) ).serialize( el )` ? (Likely with some caching of a serializer)

Reply:
Thanks Andrew but unfortunately it's not working in my testing, we more likely have to specify the same rules used by TinyMCE internally

Reply:
> Is it a requirement since we're not using this when getting the whole content of the editor but only when splitting?

Well, the content will end up getting passed to the store, so I guess so. On consecutive edits we may pass `getContent` again, but this doesn't necessarily happen. Say you split a paragraph in two and the first part contains some elements with internal attributes. This will be passed on eventually to (block) `setAttributes`.

Not sure immediately how to fix either but I guess @aduth goes in the right direction. Maybe parse and serialise if you want to work on the string? I'll have a look if we can work on the fragment.

Reply:
Did a little test here: https://github.com/WordPress/gutenberg/commit/39d9966a624eb625cd5ffc7cd6db433424bf73da

Splitting is currently entirely broken with the HTML format. The diff fixes it and should also get rid of internal stuff.

Reply:
After the splitting fix, but without serialising, you'd get this:

```
The goal of this new editor is to make adding rich content to WordPress simple and enjoyable. This whole post is composed of <em data-mce-selected="inline-boundary">pieces of</em>
```

when splitting the first paragraph of the demo content.